### ISSUE TITLE
Question about "loss is nan during training" and "bad weights, reinitializing".

### ISSUE THREAD

**Issue by Jia-Wei-Liu** [apple/ml-neuman#9]
Thanks for releasing this excellent work!

I am trying to reproduce the results in the bike scene following the 2-stage training strategy. However, in the 2nd stage, I find that sometimes the total training loss is nan and it reports "loss is nan during training" (I attach the complete report problem below). I wonder is it common to encounter such problem or am I doing something wrong? And why the loss would be nan during training and how to avoid it?
Another question is that I sometimes encounter "bad weights, reinitializing" problem. Could you please give me some hints on where I might be wrong and how to avoid such problem? Thank you very much!

"loss is nan during training {'fine_rgb_loss': nan, 'lpips_loss': 0.0011485698632895947, 'color_range_reg': nan, 'smpl_sym_reg': 6.46794360363856e-05, 'smpl_shape_reg': 0.00019673457427415997, 'mask_loss': 2.684376340766903e-05, 'sparsity_reg': 0.00018622333300299942, 'rgb_loss': nan, 'can_loss': 0.00026141401031054556, 'total_loss': nan}"

Best regards!

**Comment by jiangwei221** [comment]
Hmmmm, I don’t recall that we had NaN loss problem before. It looks like the RGB branch of the human NeRF is predicting super large values? Can you print some predicted RGB values?
As for the “bad weights” it usually shows up in the beginning of the training, if you encountered it after several thousands iteration, it means the NeRF model collapsed.
Can you also post some tensorboard screenshots?

**Comment by jiangwei221** [comment]
Why LPIPS loss is not NaN though…

**Comment by Jia-Wei-Liu** [comment]
Thanks for your quick reply!

The tensorboard does not record such nan values for all kinds of losses. But during training, sometimes it report "loss is nan during training" (about 50k in the following image) in the terminal. I am also a bit confused about this, and also confused why LPIPS loss is not nan..

<img width="827" alt="image" src="https://user-images.githubusercontent.com/52621306/185782172-2b2563c4-8c7e-4e74-8107-19485f15b17c.png">

**Comment by jiangwei221** [comment]
The loss curves are similar to ours. Could you also share some validation renderings? Is the canonical human looks ok?

**Comment by Jia-Wei-Liu** [comment]
Thanks for the quick reply! Yes, the canonical human looks okay.
This nan loss problem happens several times during training (I notice this problem when I check the terminal outputs occasionally during training). I checked the code (line557-line579 in human_nerf_trainer.py), and it seems the tensorboard will not record the losses if the total loss is nan, and the nan loss will not be back propagated. I think that's why tensorboard looks okay and the rendering is also okay. But I am still a bit confused why it shows some nan result during training (maybe I am doing something wrong), and why the training can get back to the right track by just not back propagating the nan losses.
Below I attached several test results rendered from 70k iters training ckpt, which looks okay.

<img width="1009" alt="image" src="https://user-images.githubusercontent.com/52621306/185800039-dee9e326-51d6-49f2-8f44-f641e576df22.png">
<img width="1006" alt="image" src="https://user-images.githubusercontent.com/52621306/185800070-e5dd9224-019d-4e82-9cdf-417a02cae40d.png">

**Comment by jiangwei221** [comment]
I’ll start a training at my end and monitor this NaN problem, and get back to you if I find anything new.

**Comment by jiangwei221** [comment]
Btw, can you also share some canonical renderings?

**Comment by Jia-Wei-Liu** [comment]
Thank you very much for your help! Sure! Below I attach several validation results at 110k iters.

<img width="910" alt="image" src="https://user-images.githubusercontent.com/52621306/185847256-bf45dc86-8a52-4d79-849d-380b20a4c53e.png">
<img width="897" alt="image" src="https://user-images.githubusercontent.com/52621306/185847296-aef483e2-fc18-478f-9698-7f5439cafd70.png">
<img width="998" alt="image" src="https://user-images.githubusercontent.com/52621306/185847314-408a1497-413c-42d6-a089-39443bdd78a5.png">
<img width="999" alt="image" src="https://user-images.githubusercontent.com/52621306/185847350-adfb3105-101c-4a05-9d91-baec8c98ca5b.png">

**Comment by jiangwei221** [comment]
I think I know why this happens. The sample perturbation(perturb) is set to `1`, and it may cause two adjacent samples to have 0 distance. This will cause the ray direction to become NaN in the following code:
https://github.com/apple/ml-neuman/blob/4e548bc1ec1385e610759da43eed4593d1b87764/trainers/human_nerf_trainer.py#L274-L277
One way to avoid this problem is to set `perturb` to 0.

**Comment by Jia-Wei-Liu** [comment]
Thank you very much for your help! This clarified all my questions!
